### PR TITLE
Getting Started on Backend

### PR DESCRIPTION
**Pull request by PatrickPLi**
Decided to use C++ for backend. Have a basic app which is able to receive matrices from frontend on port 45067 and return a result.

### CODE REVIEW

File: backend/main.cpp [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+void handle_options_request(ip::tcp::socket& socket) {
+    // Send CORS headers for preflight request
+    std::string response = "HTTP/1.1 200 OK\r\n";
+    response += "Access-Control-Allow-Origin: *\r\n";
+    response += "Access-Control-Allow-Methods: DELETE, POST, GET, OPTIONS\r\n";
+    response += "Access-Control-Allow-Headers: Content-Type, Authorization, X-Requested-With\r\n";
+    response += "Content-Length: 0\r\n\r\n";
+    boost::asio::write(socket, boost::asio::buffer(response));
+}
+
```
Comment (Jordandde) [inline backend/main.cpp:30]: Ideally, once we have a prototype and stuff we can move all this code out of main into it's own api file/package akin to how I did it in Go

File: backend/main.cpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            // Send a 405 Method Not Allowed response
+            std::string response = "HTTP/1.1 405 Method Not Allowed\r\n";
+            response += "Content-Type: text/plain\r\n";
+            response += "Access-Control-Allow-Origin: *\r\n"; // Enable CORS for all origins
+            response += "Content-Length: 0\r\n\r\n";
+            boost::asio::write(socket, boost::asio::buffer(response));
+            return;
+        }
+
+        if (request.empty()) {
```
Comment (Jordandde) [inline backend/main.cpp:51]: I dont think this line is reachable with an empty request, cause req == "" != "POST", I dont actually care if we handle it cause it's just gonna be us sending api requests, but just making notes if we want to organize the code later

File: backend/main.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        // CONVERT FLOATS TO FPGA/HARDWARE FRIENDLY FORMAT, PROBABLY INTS??
+        // TALK WITH FPGA AND GET RESULTS
+        // SCALE RESULTS "12-bit dac with 4.096 reference voltage -> analog switch with 15ohms of resistance 
+        // -> analog multipler that divides the result by a factor of 10 -> 16-bit adc with 4.096 reference voltage"
+
+        // UPDATE RESULT MATRIX
+        // TODO: REMOVE THIS DUMMY OPERATION
+        // Multiply matrices 1 and 2 and store the result in the result matrix
+        // This is a leetcode question I think LOL
+        for (int i = 0; i < matrices.input_matrix_1.get_num_cols(); i++) {
```
Comment (Jordandde) [inline backend/main.cpp:100]: Note for me to implement later: Add compiler directives eventually so we can leave these things in as debug operations

File: backend/main.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        // Print result matrix for debugging
+        std::cout << "Result Matrix:" << std::endl;
+        for (const auto& row : matrices.result_matrix) {
+            for (const auto& cell : row) {
+                printf("%.2f ", cell);
+            }
+            std::cout << std::endl;
+        }
+
+        // AI did these, I have no idea if this is legit -Pat
```
Comment (Jordandde) [inline backend/main.cpp:117]: yuh its legit

File: backend/main.cpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            std::string request;
+            boost::system::error_code error;
+            size_t bytes_transferred;
+            do {
+                char buffer[1024]; // Adjust buffer size as needed
+                bytes_transferred = socket.read_some(boost::asio::buffer(buffer), error);
+                request.append(buffer, buffer + bytes_transferred);
+            } while (0);
+            // Idk why but the original thing chatgpt suggested was this with no loop and the requests would just be zero length?? 
+            // Then it suggested do-while(bytes_transferred > 0) but that would just hang forever so I just put 0 and it works fine
```
Comment (Jordandde) [inline backend/main.cpp:175]: this is the same as not having a loop tho, did it not work without a loop?

**Review by Jordandde — APPROVED** [review]
🚢, none of my comments have any functional significance so its gucci

File: backend/main.cpp (hunk 5 above)
Comment (PatrickPLi) [reply backend/main.cpp:175]: I looked more closely, and I think the actual difference was that socket.read() didn’t work but socket.read_some() did... I'll play with it some more later but for now it works

**Pull request merged**